### PR TITLE
Feat!: templated android config

### LINKED ISSUE #1768: [FEATURE] Automate android build settings
**What?**
Create a workflow to automate updating the android build settings for a deployment. 

This would automate updating the "App ID and Name" manually in various places, listed in the issue https://github.com/IDEMSInternational/parenting-app-ui/issues/1237. 

**Additional comment**
First use case (which we can use to test this) could be the WASH App deployment, which currently only has a web preview but will need a Play Store App by the start of Mar 2023.

### LINKED ISSUE #2119: [FEAT] New workflow: populate constants to build files
**What?**
Various constants in the Capacitor/Android build files need to be updated with deployment specific values when building an Android bundle/apk for a given app deployment (see issue #1237). A workflow should be added that makes the appropriate changes to the files, taking the source values from the deployment config where possible.

**Why?**
This has been desirable for a long time in order to avoid the manual process, and is more easily enabled by #1210. But there is a need for the feature now that the code and content repos are separate: these files, which live in the code repo, should not be changed to match a particular release.

**How?**
Making a simple string substitution in the relevant files seems like an acceptable implementation. For an impending release of the `plh_facilitator_my` deployment, the `android-build` workflow action was updated on that deployment repo in order to handle substituting the strings, see the updated file [here](https://github.com/IDEMSInternational/plh-facilitator-app-my-content/blob/main/.github/workflows/android_build.yml) (specifically the steps `Update config names in android files` and `Add version number to build.gradle`). Whilst this action could be extended to all deployment repos in its current form, the target strings in the build files should be replaced with placeholders that do not relate to any particular deployment, and the action updated accordingly. At that point, in order to build an app bundle locally, it would be sensible to have a local workflow that could be run to populate the relevant files based on the currently active deployment. This local workflow, e.g. `yarn workflow populate_build_constants` could then be called in the deployment repo github action, replacing the explicit logic that is [there currently](https://github.com/IDEMSInternational/plh-facilitator-app-my-content/blob/main/.github/workflows/android_build.yml#L113).

**Other details**
It would be sensible to include separate workflows, with a dedicated workflow for updating the version number (as in #2070 for example). This could be called when making code releases (#2112)

### LINKED ISSUE #2136: [FEATURE] Deployment repo android build improvements
**What?**
When creating android releases on deployment branch main app repo is checked out during action, and therefore has the default `capacitor.config.ts`. It would be useful to allow authors to specify things like app_id and other config overrides in the deployment config.ts, and include overwrite as part of android build workflows

Additionally the android versioning script works from the main package.json, but would probably make more sense to assign based on content version (or combined content/builder)

Ideally these updates can then be incorporated into a new template in `packages/actions/templates`

It would be worth discussing any other requirements for content repos or how best to prioritise/achieve these items

**Other details**
Workarounds for issue above in deployment repo:
https://github.com/IDEMSInternational/plh-facilitator-app-my-content/blob/main/.github/workflows/android_build.yml

### PR DESCRIPTION
**Pull request by chrismclarke**
PR Checklist

- [ ] PR title descriptive (can be used in release notes)

## Breaking Changes
- Once merged any content repos targeting android build must have `app_id` and `app_name` specified in their config.ts android configuration

## Description
Refactor android configuration to allow any deployment to be built or run as an android app locally, and provide a template for future ios deployment-based configuration.

Existing support for building deployment apps has been refactored to work in the same way as local development, removing a lot of the existing hardcoded configurations and shell scripting requirements. 

The main changes by feature are:

#### Deployment Config
- Add support for android `app_id` and `app_name` variables. It is proposed to include these in the config themselves as supplying as part of github actions only makes it harder to test/develop locally.

#### Workflows
- Add new `android configure` workflow which handles android config file replacements
- Add new `android` workflow as a convenience to call configure workflow alongside splash and launcher icon workflows
- Remove `populate_android_assets` workflow, but the 2 child workflows it calls still accessible within the android namespace, i.e. `yarn workflow android set_splash_image` and `yarn workflow android set_launcher_icon` 
- Documentation updated where workflow references found

#### Github Actions
reusable-android-build.yml - no longer required `APP_ID` and `APP_NAME` variables, android config shell scripts removed
reusable-app-build.yml - no longer required `APP_ID` and `APP_NAME` variables, android config shell scripts removed
reusable-content-sync.yml - versioning now handled by script

#### Android Folder
- Add a new `templates` subfolder used to locate specific templates that need to be moved after populating (templated `strings.xml` conflicts with generated and templated `MainActivity.java` needs to sit in subfolder specific to app id)

#### Other Changes
- env-replace output log added
- add template files for core android build files
- `yarn scripts version` update to no longer populate `--android` version code (now read from `content_tag_latest` when setting up android build). Also added support for `--content` to directly update the `content_tag_latest` deployment config, and a `--auto-patch` to automatically increment version code patch by 1 (e.g. `1.2.3` -> `1.2.4`). 
This reduces complexity needed to version in github actions and catches an error in action code where `config.git.content_tag_latest = xxx` would not be supported (only `config.git = {content_tag_latest: xxx}`)

## Review Notes
**Android configuration**
- Update deployment config to include android app_id and app_name, e.g.
```
config.android={
  app_id:'international.idems.debug_app',
  app_name:'Debug App',
}
```
- Run new workflow
```
yarn workflow android
```
View changes made to android build files (and confirm they are as expected). 
If testing on android ensure you have a google-services.json file locally which includes entry for the target appId

Note this will also run the splash and logo workflows, to just run the config can instead use `yarn workflow android configure`

**Content Versioning**
Prompt config.ts content version update via
```sh
yarn scripts version --content
```

Auto-update config.ts content version via
```sh
yarn scripts version --content --auto-patch
```

**Github Actions**
Not easy to test, ideally we should create a debug_app play store entry which can be used to test from the debug_app content repo once merged

## Follow-up tasks
- [ ] Update relevant documentation for new android app config (@jfmcquade)... See old notes in #1237
- [ ] Generate similar workflow as required for ios (@jfmcquade)
- [ ] Remove all template-generated files and add to gitignore (currently retained to more easily view diffs) (@chrismclarke )
- [ ] Consider refactoring `reusable-android-release.yml` so that no workflows need `APP_ID` variable provided (can either load from deployment, or use previous workflow to write to output/disk and retrieve) (@ChrisMarsh82 or @chrismclarke  )

## Git Issues

Closes #2136
Closes #2119
Closes #1768

## Screenshots/Videos
**Example android workflow variable replacements**
(splash fails on debug, but rest of files configured)
image

**Example running debug_app independently on device (previous used plh_teens)**
image

image

### CODE REVIEW

**Review by jfmcquade — APPROVED** [review]
Looking good. The workflows make a lot of sense, and nice how they clean up the github actions. Good to finally be closing those issues.

Should this PR be marked as breaking, in that `config.android.app_id` and `config.android.app_name` need to be set on a deployment before an Android build can be made?

I've tested `yarn workflow android`, and everything looks how it should.

Screenshot 2024-03-22 at 15 07 22

I've tested the new flags for `yarn scripts version`. Everything seems to update as it should.

In terms of follow-ups:
- I can update the documentation for the new config
- I can create the equivalent iOS workflow (#2238 and #2239), it will give me a chance to understand the code more deeply

File: packages/scripts/src/commands/version.ts [hunk 1]
```diff
@@ -73,30 +97,31 @@ async function promptVersion(prefix: string, previousVersion: string) {
...
+    console.log(tsPath);
+    Logger.error({
+      msg1: "Existing content version not detected, Please update config.ts with:",
+      msg2: `git.content_tag_latest = ${newVersion}`,
+    });
+  }
+  // use regex to capture content_tag both inline and nested. E.g.
+  // `git: {config_tag_latest: "1.2.3"}` or `git.config_tag_latest = 1.2.3`
+  // and keep existing formatting (just replace matched version code between quotation marks)
+  const regex = /(content_tag_latest.*)"(.*)"/g;
```
[the remark this answers is not in the document]
Comment (chrismclarke) [reply packages/scripts/src/commands/version.ts]: Good idea, the versioning script will always set using double quote but no harm catching. Updated with 007f3b7b23b05c02a52b5a7570f4e9dd90d7f48f

File: packages/scripts/src/tasks/providers/android.ts [hunk 2]
```diff
@@ -1,8 +1,74 @@
...
+    // include both android folder and root (capacitor.config.ts)
+    includeFolders: ["android/**", "."],
+    envAdditional: {
+      APP_ID: appId,
+      APP_NAME: appName,
+      VERSION_CODE: versionCode,
+      VERSION_NAME: versionName,
+    },
+    // do not overwrite ${applicationId} variable in AndroidManifest template
+    excludeVariables: ["applicationId"],
```
[the remark this answers is not in the document]
Comment (chrismclarke) [reply packages/scripts/src/tasks/providers/android.ts:47]: Yeah was disappointed not to be the first to think of templating variables...
Our variable regex pattern for string inside `${...}` is case-insensitive `[a-z0-9_]`, so does need to explicitly exclude (thought about restricting upper-case but also pretty common place so doesn't achieve much).

When working on IOS i can see there's a bunch of variables in `Info.plist` in the form `$(PRODUCT_NAME)` which should not get matched, but then things like `${PODS_PODFILE_DIR_PATH}` in `project.pbxproj` that would. Given the project file has all the xcode config which also defines things like deployment targets, will likely either want to exclude that file from the replace action or manually set `includeVariables` to only define those you want to swap (as exclusion list could be a little long and confusing)

**Review by esmeetewinkel — APPROVED** [review]
Functional test passed on `debug` and `plh_facilitator_mx` deployments
image
image

> ideally we should create a debug_app play store entry which can be used to test from the debug_app content repo once merged

I agree that having a debug Play Store app could be useful. I had assumed we never set this up because it wouldn't make sense to fill out all Google's declarations about data, privacy policy, target audience etc for the debug app. But thinking about it maybe that's only required when going beyond the "internal testing" stage... Happy to set this up

**Pull request merged**